### PR TITLE
20913/20915: Add transit solution for Azure China and AWS China

### PR DESCRIPTION
**Pull request by mlin-aviatrix**

### CODE REVIEW

File: aviatrix/data_source_aviatrix_gateway.go [hunk 1]
```diff
@@ -505,19 +505,19 @@ func dataSourceAviatrixGatewayRead(d *schema.ResourceData, meta interface{}) err
...
 				if gwHaGw.InsaneMode == "yes" {
 					d.Set("peering_ha_insane_mode_az", gwHaGw.GatewayZone)
 				}
-			} else if gwHaGw.CloudType == goaviatrix.AZURE || gwHaGw.CloudType == goaviatrix.OCI {
+			} else if goaviatrix.IsCloudType(gwHaGw.CloudType, goaviatrix.AZURECHINA|goaviatrix.OCIRelatedCloudTypes) {
 				d.Set("peering_ha_subnet", gwHaGw.VpcNet)
-			} else if gwHaGw.CloudType == goaviatrix.GCP {
+			} else if goaviatrix.IsCloudType(gwHaGw.CloudType, goaviatrix.GCPRelatedCloudTypes) {
 				d.Set("peering_ha_zone", gwHaGw.GatewayZone)
 			}
```
Comment (CyrusJavan) [inline aviatrix/data_source_aviatrix_gateway.go]: What happens if cloud_type == AZURE? I don't see a branch for AZURE anymore

Comment (mlin-aviatrix) [reply aviatrix/data_source_aviatrix_gateway.go]: It should be all azure clouds not just azure china.

File: aviatrix/resource_aviatrix_transit_gateway.go [hunk 2]
```diff
@@ -521,7 +526,7 @@ func resourceAviatrixTransitGatewayCreate(d *schema.ResourceData, meta interface
 
 	cloudType := d.Get("cloud_type").(int)
 	zone := d.Get("zone").(string)
-	if cloudType != goaviatrix.AZURE && zone != "" {
+	if goaviatrix.IsCloudType(cloudType, goaviatrix.AZURE) && zone != "" {
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_transit_gateway.go]: Should be `!goaviatrix.IsCloudType(cloudType, goaviatrix.AZURE)`

File: aviatrix/resource_aviatrix_transit_gateway.go [hunk 3]
```diff
@@ -623,33 +628,33 @@ func resourceAviatrixTransitGatewayCreate(d *schema.ResourceData, meta interface
 	lanVpcID := d.Get("lan_vpc_id").(string)
 	lanPrivateSubnet := d.Get("lan_private_subnet").(string)
 	if enableTransitFireNet {
-		if !intInSlice(gateway.CloudType, []int{goaviatrix.AWS, goaviatrix.AWSGOV, goaviatrix.GCP, goaviatrix.AZURE}) {
-			return fmt.Errorf("'enable_transit_firenet' is only supported in AWS, AWSGOV, GCP and AZURE providers")
+		if !goaviatrix.IsCloudType(cloudType, goaviatrix.AWSRelatedCloudTypes|goaviatrix.GCPRelatedCloudTypes|goaviatrix.AzureArmRelatedCloudTypes) {
+			return fmt.Errorf("'enable_transit_firenet' is only supported in AWS (1), GCP (4), AZURE (8), AWSGOV (256), AWSCHINA (1024) and AZURECHINA (2048)")
 		}
-		if intInSlice(gateway.CloudType, []int{goaviatrix.AZURE, goaviatrix.GCP}) {
+		if goaviatrix.IsCloudType(cloudType, goaviatrix.GCPRelatedCloudTypes) || goaviatrix.IsCloudType(cloudType, goaviatrix.AzureArmRelatedCloudTypes) {
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_transit_gateway.go]: Why not just combine the int with `|` in a single call to IsCloudType?

File: aviatrix/resource_aviatrix_transit_gateway.go [hunk 4]
```diff
@@ -1629,14 +1644,14 @@ func resourceAviatrixTransitGatewayUpdate(d *schema.ResourceData, meta interface
...
+	if d.HasChange("enable_transit_firenet") && goaviatrix.IsCloudType(gateway.CloudType, goaviatrix.GCPRelatedCloudTypes|goaviatrix.AzureArmRelatedCloudTypes) {
 		return fmt.Errorf("editing 'enable_transit_firenet' in GCP and AZURE is not supported")
 	}
 	if d.Get("enable_egress_transit_firenet").(bool) && !d.Get("enable_transit_firenet").(bool) {
 		return fmt.Errorf("'enable_egress_transit_firenet' requires 'enable_transit_firenet' to be set to true")
 	}
-	if d.Get("enable_egress_transit_firenet").(bool) && !intInSlice(gateway.CloudType, []int{goaviatrix.AWS, goaviatrix.AZURE, goaviatrix.AWSGOV, goaviatrix.GCP}) {
-		return fmt.Errorf("'enable_egress_transit_firenet' is currently only supported on AWS, AZURE, GCP and AWSGOV cloud providers")
+	if d.Get("enable_egress_transit_firenet").(bool) && !goaviatrix.IsCloudType(gateway.CloudType, goaviatrix.AWSRelatedCloudTypes|goaviatrix.GCPRelatedCloudTypes|goaviatrix.AzureArmRelatedCloudTypes^goaviatrix.AZURECHINA) {
+		return fmt.Errorf("'enable_egress_transit_firenet' is currently only supported on AWS (1), GCP (4), AZURE (8), AWSGOV (256), AWSCHINA (1024) and AZURECHINA (2048)")
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_transit_gateway.go]: Not supported on AZURECHINA right?

File: aviatrix/resource_aviatrix_transit_gateway.go [hunk 5]
```diff
@@ -1710,9 +1725,9 @@ func resourceAviatrixTransitGatewayUpdate(d *schema.ResourceData, meta interface
 		oldZone, newZone := d.GetChange("ha_zone")
 		deleteHaGw := false
 		changeHaGw := false
-		if intInSlice(transitGw.CloudType, []int{goaviatrix.AWS, goaviatrix.AZURE, goaviatrix.AWSGOV, goaviatrix.ALIYUN, goaviatrix.OCI}) {
+		if goaviatrix.IsCloudType(transitGw.CloudType, goaviatrix.AWSRelatedCloudTypes|goaviatrix.AzureArmRelatedCloudTypes|goaviatrix.AliyunRelatedCloudTypes) {
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_transit_gateway.go]: Also OCI

File: aviatrix/resource_aviatrix_transit_gateway.go [hunk 6]
```diff
@@ -2079,7 +2094,8 @@ func resourceAviatrixTransitGatewayUpdate(d *schema.ResourceData, meta interface
 	if enableGatewayLoadBalancer && !enableFireNet && !enableTransitFireNet {
 		return fmt.Errorf("'enable_gateway_load_balancer' is only valid when 'enable_firenet' or 'enable_transit_firenet' is set to true")
 	}
-	if enableGatewayLoadBalancer && gateway.CloudType != goaviatrix.AWS {
+	// Enable gateway load balancer is not supported in AWSCHINA for 6.4
+	if enableGatewayLoadBalancer && goaviatrix.IsCloudType(gateway.CloudType, goaviatrix.AWSRelatedCloudTypes^goaviatrix.AWSCHINA) {
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_transit_gateway.go]: Can you verify with spencer, I think also AWSGOV does not support gateway load balancer

File: aviatrix/resource_aviatrix_vpc.go [hunk 7]
```diff
@@ -359,7 +359,7 @@ func resourceAviatrixVpcRead(d *schema.ResourceData, meta interface{}) error {
 	d.Set("cloud_type", vC.CloudType)
 	d.Set("account_name", vC.AccountName)
 	d.Set("name", vC.Name)
-	if vC.CloudType != goaviatrix.GCP {
+	if goaviatrix.IsCloudType(vC.CloudType, goaviatrix.GCPRelatedCloudTypes) {
```
Comment (CyrusJavan) [inline aviatrix/resource_aviatrix_vpc.go]: Should be` !goaviatrix.IsCloudType(vC.CloudType, goaviatrix.GCPRelatedCloudTypes)`

File: docs/resources/aviatrix_spoke_gateway.md [hunk 8]
```diff
@@ -96,6 +96,40 @@ resource "aviatrix_spoke_gateway" "test_spoke_gateway_awsgov" {
...
+```
+```hcl
+# Create an Aviatrix Azure China Spoke Gateway
+resource "aviatrix_spoke_gateway" "test_spoke_gateway_azure" {
+  cloud_type                        = 2048
+  account_name                      = "my-azure-china"
+  gw_name                           = "spoke-gw-01"
+  vpc_id                            = "spoke:test-spoke-gw-123"
+  vpc_reg                           = "China North"
+  gw_size                           = "Standard_A0"
```
Comment (CyrusJavan) [inline docs/resources/aviatrix_spoke_gateway.md:125]: Needs to have `storage_name` right?

Comment (CyrusJavan) [reply docs/resources/aviatrix_spoke_gateway.md:125]: Also should have storage_name below with a description

File: docs/resources/aviatrix_transit_gateway.md [hunk 9]
```diff
@@ -104,6 +104,43 @@ resource "aviatrix_transit_gateway" "test_transit_gateway_awsgov" {
...
+resource "aviatrix_transit_gateway" "test_transit_gateway_azure_china" {
+  cloud_type         = 2048
+  account_name       = "devops_azure_china"
+  gw_name            = "transit"
+  vpc_id             = "vnet1:hello"
+  vpc_reg            = "China North"
+  gw_size            = "Standard_A0"
+  subnet             = "10.30.0.0/24"
+  ha_subnet          = "10.30.0.0/24"
+  ha_zone            = "az-2"
```
Comment (CyrusJavan) [inline docs/resources/aviatrix_transit_gateway.md:138]: storage_name?

File: docs/resources/aviatrix_transit_gateway.md [hunk 10]
```diff
@@ -232,12 +269,15 @@ The following arguments are supported:
...
+* `tags` - (Optional) Map of tags to assign to the gateway. Only available for AWS, Azure, AWSGov, AWSChina and AzureChina gateways. Allowed characters vary by cloud type but always include: letters, spaces, and numbers. AWS, AWSGov and AWSChina allow the following special characters: + - = . _ : / @.  Azure and AzureChina allows the following special characters: + - = . _ : @. Example: {"key1" = "value1", "key2" = "value2"}.
 * `tunnel_detection_time` - (Optional) The IPsec tunnel down detection time for the Transit Gateway in seconds. Must be a number in the range [20-600]. The default value is set by the controller (60 seconds if nothing has been changed). **NOTE: The controller UI has an option to set the tunnel detection time for all gateways. To achieve the same functionality in Terraform, use the same TF_VAR to manage the tunnel detection time for all gateways.** Available in provider R2.19+.
+* `storage_name` (Optional) Specify a storage account. Require if `cloud_type` is 2048 (AZURECHINA). Available as of Provider version 2.19+.
+
+>>>>>>> 20913/20915: Add transit solution for Azure China and AWS China
```
Comment (CyrusJavan) [inline docs/resources/aviatrix_transit_gateway.md]: clean up

**Review by CyrusJavan — CHANGES_REQUESTED** [review]

**Review by CyrusJavan — APPROVED** [review]
looks good to me

**Review by shuoliugit — APPROVED** [review]

**Review by zhiqiangjin — APPROVED** [review]

**Pull request merged**